src/Nethermind/Nethermind.Shutter/ShutterApi.cs

```diff
@@ -0,0 +1,154 @@
+    public ShutterBlockImprovementContextFactory? BlockImprovementContextFactory;
+
+    protected readonly TimeSpan _slotLength;
+    protected readonly TimeSpan _blockUpToDateCutoff;
+    protected readonly IReadOnlyBlockTree _readOnlyBlockTree;
+    protected readonly IBlockTree _blockTree;
+    private readonly ReadOnlyTxProcessingEnvFactory _txProcessingEnvFactory;
+    private readonly IAbiEncoder _abiEncoder;
+    private readonly ILogManager _logManager;
+    private readonly IShutterConfig _cfg;
+
+    private readonly TimeSpan _blockWaitCutoff;
+
+    public ShutterApi(
+        IAbiEncoder abiEncoder,
+        IBlockTree blockTree,
+        IEthereumEcdsa ecdsa,
+        ILogFinder logFinder,
+        IReceiptFinder receiptFinder,
+        ILogManager logManager,
+        ISpecProvider specProvider,
+        ITimestamper timestamper,
+        IWorldStateManager worldStateManager,
+        IShutterConfig cfg,
+        Dictionary<ulong, byte[]> validatorsInfo,
+        TimeSpan slotLength
+        )
+    {
+        _cfg = cfg;
+        _blockTree = blockTree;
+        _readOnlyBlockTree = blockTree.AsReadOnly();
+        _abiEncoder = abiEncoder;
+        _logManager = logManager;
+        _slotLength = slotLength;
+        _blockUpToDateCutoff = slotLength;
+        _blockWaitCutoff = _slotLength / 3;
+
+        _txProcessingEnvFactory = new(worldStateManager, blockTree, specProvider, logManager);
+
+        Time = InitTime(specProvider, timestamper);
+        TxLoader = new(logFinder, _cfg, Time, specProvider, ecdsa, abiEncoder, logManager);
+        Eon = InitEon();
+        BlockHandler = new ShutterBlockHandler(
+            specProvider.ChainId,
+            _cfg,
+            _txProcessingEnvFactory,
+            blockTree,
+            abiEncoder,
+            receiptFinder,
+            validatorsInfo,
+            Eon,
+            TxLoader,
+            Time,
+            logManager,
+            _slotLength,
+            BlockWaitCutoff);
+
+        TxSource = new ShutterTxSource(TxLoader, _cfg, Time, logManager);
+
+        KeyValidator = new ShutterKeyValidator(_cfg, Eon, logManager);
+
+        InitP2P(_cfg, logManager);
+    }
+
+    public void StartP2P(CancellationTokenSource? cancellationTokenSource = null)
+        => P2P!.Start(cancellationTokenSource);
+
+    public ShutterBlockImprovementContextFactory GetBlockImprovementContextFactory(IBlockProducer blockProducer)
+    {
+        BlockImprovementContextFactory ??= new(
+            blockProducer,
+            TxSource,
+            _cfg,
+            Time,
+            _logManager,
+            _slotLength
+        );
+        return BlockImprovementContextFactory;
+    }
+
+    public async ValueTask DisposeAsync()
+    {
+        TxSource.Dispose();
+        BlockHandler.Dispose();
+        await (P2P?.DisposeAsync() ?? default);
+    }
+
+    protected virtual async void OnKeysReceived(object? sender, IShutterP2P.KeysReceivedArgs keysReceivedArgs)
```

Comment:
Why is that?

Reply:
When you hit the first await it will return control to the one who raised the event, who might be expecting the event to run to completion since its a void method. So it can result in unexpected behavior, but it can also be fine, as in this case :)

Reply:
Nice catch @LukaszRozmej